File: src/main/java/de/logicline/fruitbagger/UserRetriever.java
package de.logicline.fruitbagger;

import de.logicline.fruitbagger.domain.FruitUser;
import io.vertx.core.Handler;
import io.vertx.core.json.JsonArray;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.auth.oauth2.AccessToken;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.Session;
import org.mongodb.morphia.Datastore;

import java.util.List;
import java.util.Optional;

public class UserRetriever implements Handler<RoutingContext> {
    // retrieve the user profile, this is a common feature but not from the official OAuth2 spec
    private final Datastore datastore;

    public UserRetriever(Datastore datastore) {
        this.datastore = datastore;
    }

    @Override
    public void handle(RoutingContext ctx) {
        Session session = ctx.session();
        AccessToken user = (AccessToken) ctx.user();
        user.userInfo(res -> {
            if (res.failed()) {
                ctx.session().destroy();
                ctx.fail(res.cause());
                ctx.next();
                return;
            }
            if (session.get("fruitUser") != null) {
                ctx.next();
                return;
            }
            final JsonObject userInfo = res.result();
            user.fetch("https://api.github.com/user/emails", res2 -> {
                if (res2.failed()) {
                    ctx.session().destroy();
                    ctx.fail(res2.cause());
                    return;
                }
                JsonArray emails = res2.result().jsonArray();
                userInfo.put("private_emails", emails);

                Optional<Object> first = emails.stream().filter(email -> {
                    JsonObject email1 = (JsonObject) email;
                    return email1.containsKey("primary") && email1.getBoolean("primary");
                }).findFirst();
                if (first.isPresent()) {
                    JsonObject primaryEmail = (JsonObject) first.get();
                    String email = primaryEmail.getString("email");
                    List<FruitUser> users = datastore.find(FruitUser.class).field("email").equalIgnoreCase(email).asList();
                    FruitUser dbuser = null;
                    if (users.isEmpty()) {
                        dbuser = new FruitUser(email);
                        datastore.save(dbuser);
                    } else
                        dbuser = users.get(0);
                    session.put("fruitUser", dbuser);
                    ctx.next();
                    return;
                }
                ctx.fail(403);
            });
        });

    }
}

File: src/main/java/de/logicline/fruitbagger/UserRetriever.kt
package de.logicline.fruitbagger

import de.logicline.fruitbagger.domain.FruitUser
import io.vertx.core.Handler
import io.vertx.core.json.JsonObject
import io.vertx.ext.auth.oauth2.AccessToken
import io.vertx.ext.web.RoutingContext
import org.mongodb.morphia.Datastore
import java.util.*

class UserRetriever(private val datastore: Datastore) : Handler<RoutingContext> {

    override fun handle(ctx: RoutingContext) {
        val session = ctx.session()
        val user = ctx.user() as AccessToken
        user.userInfo { res ->
            if (res.failed()) {
                ctx.session().destroy()
                ctx.fail(res.cause())
                ctx.next()
                return@userInfo
            }
            // if the user already has registered for this browser session
            if (session.get<Any>("fruitUser") != null) {
                ctx.next()
                return@userInfo
            }
            registerUser(user, ctx)
        }

    }

    private fun registerUser(user: AccessToken, ctx: RoutingContext) {
        user.fetch("https://api.github.com/user/emails") { res ->
            if (res.failed()) {
                ctx.session().destroy()
                ctx.fail(res.cause())
                return@fetch
            }
            val emails = res.result().jsonArray()
            val getPrimaryUserMail = emails.stream().filter { email ->
                val email1 = email as JsonObject
                email1.containsKey("primary") && email1.getBoolean("primary")!!
            }.findFirst()

            if (getPrimaryUserMail.isPresent) {
                addUserToDb(getPrimaryUserMail, ctx)
                return@fetch
            }
            ctx.fail(403)
        }
    }

    private fun addUserToDb(getPrimaryUserMail: Optional<Any>, ctx: RoutingContext) {
        val primaryEmail = getPrimaryUserMail.get() as JsonObject
        val email = primaryEmail.getString("email")
        val users = datastore.find(FruitUser::class.java).field("email").equalIgnoreCase(email).asList()
        var dbuser: FruitUser?
        if (users.isEmpty()) {
            dbuser = FruitUser(email)
            datastore.save(dbuser)
        } else
            dbuser = users[0]
        ctx.session().put("fruitUser", dbuser)
        ctx.next()
    }
}
